pkg/syncer/volume_health.go

```diff
@@ -77,58 +77,75 @@ func csiGetVolumeHealthStatus(ctx context.Context, k8sclient clientset.Interface
 		}
 	}
 
-	for _, vol := range queryResult.Volumes {
-		log.Debugf("Volume %q Health Status %q", vol.VolumeId.Id, vol.HealthStatus)
+	for volID, pvc := range volumeHandleToPvcMap {
+		foundVol := false
+		for _, vol := range queryResult.Volumes {
+			log.Debugf("Volume %q Health Status %q", vol.VolumeId.Id, vol.HealthStatus)
+			if volID == vol.VolumeId.Id {
+				log.Debugf("csiGetVolumeHealthStatus: Found pvc %q for volume %q", pvc, vol.VolumeId.Id)
+				foundVol = true
+				// only update PVC health annotation if the HealthStatus of volume is not "unknown"
+				if vol.HealthStatus != string(pbmtypes.PbmHealthStatusForEntityUnknown) {
+					volHealthStatus, err := common.ConvertVolumeHealthStatus(vol.HealthStatus)
+					if err != nil {
+						log.Errorf("csiGetVolumeHealthStatus: invalid health status %q for volume %q", vol.HealthStatus, vol.VolumeId.Id)
+					}
+					updateVolumeHealthStatus(ctx, k8sclient, pvc, volHealthStatus)
+				}
+			}
+		}
+		if !foundVol {
+			// Set volume health status as "Inaccessible" when PVC is not found in CNS.
+			// When a Datastore is removed from VC (like vSAN direct disk decommisson with noAction does), the CNS Volumes
```

Comment:
CNS DB is a cache and eventually consistent. Its quite possible that a valid accessible volume may not be in the CNS cache DB yet. So, is it fine for such volumes to also show up as inaccessible? Cc @subramanian-neelakantan

Reply:
For those volumes, the volume health status will be set back to accessible during the next fullsync cycle since it will be shown up in CNS.

Reply:
Sure, is that behavior acceptable to vDPp operators?

Reply:
@subramanian-neelakantan has requested this change.  Let's have him confirm.

Reply:
(Ref https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/852/files#r631577836), since we are taking the design choice of reflecting the CNS cached status of health on the PVC's health annotation at any given point of time, this change here is correct. The vDPp operators are advised to look at the health change timestamp and wait "long enough" (like an hour) before taking any corrective actions. So if the PVC health is getting updated every 5 mins,  waiting for an hour or so, makes this an acceptable level of eventual consistency.

Reply:
Ok, lets capture this as a comment in this file and also in the CSI docs. @xing-yang can we track the CSI doc update as a separate issue/bug so that we do not forget it?

Reply:
@SandeepPissay I can open an issue on Github to track the CSI doc issue.  This is only for Supervisor/Guest cluster.

Reply:
Opened an issue for doc here: https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/955

Reply:
Thanks!